From a677c93994d961813b8e2689154a258a65fed667 Mon Sep 17 00:00:00 2001
From: Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>
Date: Sat, 19 Apr 2025 20:37:48 +0200
Subject: [PATCH] kerndat: add madvise(MADV_GUARD_INSTALL) feature-detection

Signed-off-by: Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>
(cherry picked from commit 2bb77daa92d26266d32e08ac21c0ed91f438a945)

Conflicts:
	criu/include/kerndat.h
	criu/kerndat.c
---
 criu/include/kerndat.h |  1 +
 criu/kerndat.c         | 32 ++++++++++++++++++++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/criu/include/kerndat.h b/criu/include/kerndat.h
index 60192fca7..881a5f1e6 100644
--- a/criu/include/kerndat.h
+++ b/criu/include/kerndat.h
@@ -83,6 +83,7 @@ struct kerndat_s {
 	bool has_rseq;
 	bool has_ptrace_get_rseq_conf;
 	bool has_pagemap_scan;
+	bool has_madv_guard;
 };
 
 extern struct kerndat_s kdat;
diff --git a/criu/kerndat.c b/criu/kerndat.c
index 19ab85c44..fbfd732b8 100644
--- a/criu/kerndat.c
+++ b/criu/kerndat.c
@@ -29,6 +29,7 @@
 #include "kerndat.h"
 #include "fs-magic.h"
 #include "mem.h"
+#include "mman.h"
 #include "common/compiler.h"
 #include "sysctl.h"
 #include "cr_options.h"
@@ -1489,6 +1490,33 @@ static int kerndat_has_nftables_concat(void)
 #endif
 }
 
+static int kerndat_has_madv_guard(void)
+{
+	void *map;
+
+	map = mmap(NULL, PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+	if (map == MAP_FAILED) {
+		pr_perror("Can't mmap a page for has_madv_guard feature test");
+		return -1;
+	}
+
+	if (madvise(map, PAGE_SIZE, MADV_GUARD_INSTALL)) {
+		if (errno != EINVAL) {
+			pr_perror("madvise failed (has_madv_guard check)");
+			goto mmap_cleanup;
+		}
+	} else {
+		kdat.has_madv_guard = true;
+	}
+
+	munmap(map, PAGE_SIZE);
+	return 0;
+
+mmap_cleanup:
+	munmap(map, PAGE_SIZE);
+	return -1;
+}
+
 /*
  * Some features depend on resource that can be dynamically changed
  * at the OS runtime. There are cases that we cannot determine the
@@ -1698,6 +1726,10 @@ int kerndat_init(void)
 		pr_err("kerndat_has_ptrace_get_rseq_conf failed when initializing kerndat.\n");
 		ret = -1;
 	}
+	if (!ret && kerndat_has_madv_guard()) {
+		pr_err("kerndat_has_madv_guard has failed when initializing kerndat.\n");
+		ret = -1;
+	}
 
 	kerndat_lsm();
 	kerndat_mmap_min_addr();
-- 
2.34.1

